### PR TITLE
web: create preview pane

### PR DESCRIPTION
**Pull request by yuindustries**
I tried tab-like buttons for "Logs" and "Preview" view options.

- In this PR, it's not obvious the view options are buttons. I'll address this in my next design pass.
- When we have more than two views, this design will have to change.

2019-04-04 - preview

## Nav Design Alternatives considered

I mocked up a design with the view options to the left of the sidebar, but it was a bit too visually obtrusive.

2019-04-04 - view options mock

### CODE REVIEW

File: web/src/PreviewPane.test.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import React from "react"
+import ReactDOM from "react-dom"
+import PreviewPane from "./PreviewPane"
+import renderer from "react-test-renderer"
+
+it("renders without crashing", () => {
+  let div = document.createElement("div")
+  Element.prototype.scrollIntoView = jest.fn()
+  ReactDOM.render(<PreviewPane endpoint="http://localhost:9000" />, div)
```
Comment (jazzdan) [inline web/src/PreviewPane.test.tsx]: @yuindustries and I talked and we should add a test here for when no endpoint is passed. The preview component should probably handle that gracefully.

**Comment by nicks** [comment]
one major bug:
when the logs pane is focused, you can scroll with the mousewheel but there's no scrollbar.
when the logs pane is not focused, you can't scroll at all 😭 . and still no scrollbar

some minor issues:
when looking at all logs, and clicking the "Preview" button, i get a "No match" error
when looking at (Tiltfile) logs, and clicking the "Preview" button, I get a blank screen
when the right sidebar is taller than the screen, there's no scrollbar. there used to be a scrollbar before this PR, so this appears to be a regression

File: web/src/HUD.scss [hunk 2]
```diff
@@ -1,7 +1,5 @@
-@import 'constants';
+@import "constants";
 
 .HUD {
-  display: flex;
-  background-color: $color-navy;
-  min-height: 100vh;
-}
\ No newline at end of file
+  height: 100vh;
```
Comment (nicks) [inline web/src/HUD.scss]: i'm pretty sure the change from min-height to height here is wrong; why would we want to make the HUD exactly 100vh?

File: web/src/HUD.tsx [hunk 3]
```diff
@@ -139,15 +157,25 @@ class HUD extends Component<HudProps, HudState> {
       combinedLog = view.Log
     }
 
+    let previewRoute = (props: RouteComponentProps<any>) => {
+      let name = props.match.params ? props.match.params.name : ""
+      let endpoint = ""
+      if (view && name !== "") {
+        let r = view.Resources.find(r => r.Name === name)
+        endpoint = r ? r.Endpoints && r.Endpoints[0] : ""
```
Comment (nicks) [inline web/src/HUD.tsx]: if there's no endpoint, we probably don't want to return a PreviewPane at all? maybe better to fallback to a LogPane?

File: web/src/LoadingScreen.scss [hunk 4]
```diff
@@ -1,8 +1,14 @@
+@import "constants";
+
 .LoadingScreen {
-  min-height: 100vh;
+  position: fixed;
```
Comment (nicks) [inline web/src/LoadingScreen.scss]: this seems wrong to me? now the loading screen is off-center and looks kind of broken

File: web/src/LogPane.scss [hunk 5]
```diff
@@ -1,16 +1,27 @@
 @import "constants.scss";
 
 .LogPane {
-  margin: $spacing-unit / 2;
+  position: fixed;
```
Comment (nicks) [inline web/src/LogPane.scss]: hmmm...i understand that using "position: fixed" is a cheap fix to force the layout to do what you want.

But in practice, putting the main content of a page inside a position:fixed container is almost always something you'll regret. It's semantically / accessibility-wise not what we want, and going to break the interaction in lots of subtle and frustrating ways.

Let's talk about this more tomorrow.

**Pull request merged**